### ISSUE TITLE
missing or hidden LADSPA descriptor

### ISSUE THREAD

**Issue by x42** [sadko4u/lsp-plugins#47]
I'm forwarding a user-report from Ardour IRC:

> Ran AV Linux updates about a week ago, now having errors in Ardour.  Anyone know the answer?  Errors are:
```
LADSPA: module "/usr/lib/ladspa/lsp-plugins-ladspa.so" has no descriptor function.
'ladspa_descriptor': /usr/lib/ladspa/lsp-plugins-ladspa.so: undefined symbol: ladspa_descriptor
```

I am not sure if this is specific to AVLinux, or if it is an upstream issue, but I thought I'll leave a note for it to not be forgotten.

**Comment by sadko4u** [comment]
Just checked my local installation. All seems to be OK. **ladspa_descriptor** function is present, LADSPA plugins are instantiating well in Ardour.

**Comment by AVLinux** [comment]
Hi,

I only packaged the LSP Plugins with the LV2 and LXVST binaries from the LSP site until KXStudio started to package them recently, the 1.1.9 packages which are included in AV Linux are from KXStudio and now include the LADSPA Plugins, I suspect the issue is from the package upgrade to the KXStudio versions.

**Comment by x42** [comment]
https://github.com/sadko4u/lsp-plugins/blob/04cc45a4aee3bd3a46ecbd0a9160d0d87f8f8a32/src/container/ladspa.cpp#L341  does indeed not have a default visibility attribute.

I suspect that @falkTX correctly compiles the plugins with hidden visibility for kxstudio and strips all not exposed symbols. This would cause the issue.

(for LV2 that's fine the plugin uses LV2_SYMBOL_EXPORT as does the lv2.h)

**Comment by sadko4u** [comment]
I believe the problem can be that I don't use extern "C" for the ladspa_descriptor() function as done in <ladspa.h> header.

**Comment by x42** [comment]
It's about symbol visibility. "extern "C" serves a different purpose. Try

```C
#ifdef _WIN32
__declspec(dllexport)
#else
__attribute__ ((visibility ("default")))
#endif
 const LADSPA_Descriptor * ladspa_descriptor(unsigned long index) 
```

**Comment by sadko4u** [comment]
Seems to be working well after adding __declspec prefix:
```
nm -g --defined-only lsp-plugins-ladspa.so
000000000015592c T _fini
00000000000c0c80 T _init
00000000000c6180 T ladspa_descriptor
```

**Comment by sadko4u** [comment]
This patch should fix the problem: 184c7d0fbb9bd255f0275fe714ccf263af96200f

**Comment by x42** [comment]
Yep looks good. Thanks for the quick fix!

**Comment by sadko4u** [comment]
Today checked export for other libraries and finished up with this commit: d78ad86ec405c10d7e7d26fc6299f7446422f9b7

**Comment by sadko4u** [comment]
Merged changes into **devel** branch: 00f748dc4a2253ded7b94bc93f0a95a3453efa80

**Comment by x42** [comment]
extern "C" in C++, is right, but for C-code, `#define LSP_LIBRARY_EXTERN  extern` is not.

**Comment by sadko4u** [comment]
Fixed: 75ac0d5c204467320c751856bdb932efbdfaa8de
Please don't close topics, I'll do it after release

**Comment by sadko4u** [comment]
Available in 1.1.10 release!
